### PR TITLE
[nccl-pg] Pass group global rank information to NCCL PG

### PR DESCRIPTION
**Pull request by pavanbalaji**
We were only passing a subset of the group creation information to the
NCCL PG.  We are specifically missing the information on which global
ranks belong to a particular PG.

This allows the NCCL PG to use this additional information for things
like better trace logging.

Test Plan:

OSS CI

Reviewers:

Subscribers:

Tasks:

Tags:

Fixes #ISSUE_NUMBER

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @kiukchung @d4l3k @lucasllc

### CODE REVIEW

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp [hunk 1]
```diff
@@ -368,11 +368,23 @@ class TORCH_API ProcessGroupNCCL : public Backend {
...
       int size,
-      c10::intrusive_ptr<Options> options = Options::create());
+      c10::intrusive_ptr<Options> options = Options::create())
+  : Backend(rank, size) {
+    DistributedBackendOptions opts;
+    opts.store = store;
+    opts.group_rank = rank;
+    opts.group_size = size;
+    /* leave default values for the remaining opts parameters */
+    ProcessGroupNCCL(opts, options);
```
Comment (Skylion007) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:386]:
```suggestion
    ProcessGroupNCCL(opts, std::move(options));
```
Doesn't c10::instrusive_ptr have a move ctor?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 2]
```diff
@@ -679,16 +679,14 @@ bool ProcessGroupNCCL::CoalescedWorkNCCL::wait(
...
-    const c10::intrusive_ptr<Store>& store,
-    int rank,
-    int size,
+    DistributedBackendOptions& opts,
     c10::intrusive_ptr<Options> options)
-    : Backend(rank, size),
-      store_(store),
+    : Backend(opts.group_rank, opts.group_size),
+      store_(opts.store),
       options_(options),
```
Comment (Skylion007) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:686]:
```suggestion
      options_(std::move(options)),
```
shouldn't this also be std::move? I am surprised clang-tidy did not catch this though.

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp [hunk 3]
```diff
@@ -372,6 +372,15 @@ class TORCH_API ProcessGroupNCCL : public Backend {
...
   // communicator. These NCCL communicators are cached and reused if possible.
   //
+  ProcessGroupNCCL(
+      DistributedBackendOptions& opts,
+      c10::intrusive_ptr<Options> options = Options::create())
+      : ProcessGroupNCCL(
+            opts.store,
+            opts.group_rank,
+            opts.group_size,
+            options) {}
```
Comment (Skylion007) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:382]:
```suggestion
            std::move(options)) {}
```

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 4]
```diff
@@ -678,6 +678,141 @@ bool ProcessGroupNCCL::CoalescedWorkNCCL::wait(
...
+      store_(opts.store),
+      options_(options),
+      ncclCommCounter_(0),
+      traceKeyStart_(getTraceStartKey("NCCL", opts.group_rank)),
+      traceKeyEnd_(getTraceEndKey("NCCL", opts.group_rank)),
+      terminateProcessGroup_(false),
+      terminateHeartbeatMonitorThread_(false),
+      collectiveDebugInfoMode_(false),
+      uid_(process_group_id++) {
+  TORCH_CHECK_WITH(
```
Comment (fduwjj) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:694]: Can we consolidate the logic with the existing constructor?

**Review by fduwjj — COMMENTED** [review]
This is indeed interesting... I am also surprised that we don't have that even now.

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 2 above)
Comment (pavanbalaji) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp]: I don't think this should be an `std::move`.  The options object could be used inside this constructor too.

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 4 above)
Comment (fduwjj) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:694]: That's what I mean, so if we can make the old point to this new constructor, that will also be great.

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 5]
```diff
@@ -678,6 +678,141 @@ bool ProcessGroupNCCL::CoalescedWorkNCCL::wait(
 
 static std::atomic<size_t> process_group_id = 0;
 
+ProcessGroupNCCL::ProcessGroupNCCL(
```
Comment (H-Huang) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:681]: What is missing from the existing nccl constructor that DistributedBackendOptions is providing?

Comment (pavanbalaji) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp]: The global ranks associated with each PG.

**Review by kwen2501 — APPROVED** [review]
LGTM.

**Pull request closed without merging**